[Spring 지하철 노선도 - 3단계] 돌범(조재성) 미션 제출합니다.

is2js opened the pull request:
안녕하십니까 범블비! 덕분에 1,2단계에서 많이 배우고, 마지막에 주신 피드백까지 반영하였습니다. 저에겐 쉽지 않은 미션이라 시간이 걸린 점 양해부탁드리겠습니다.! 
다시 잘 부탁드리겠습니다!

### 1,2단계 최종 피드백 반영사항
1. 부적절한 Optional 사용 수정
 - rowMapper가 null 반환하지 않는 한, null응답의 가능성이 없는 queryForObject에 대해 Opiotnal.of()로 대체 후 suppressWarning Method
2. Line, Station 도메인에 대한 name 검증 추가 -> 중복되어 Name이라는 새로운 도메인에 검증 로직 위임
   - null 검증은 requestDto의 각 필드들에 대해 검증 로직 추가
4. repository가 아닌 dao에 한하여 delete(entity) -> deleteById(id)를 사용하도록 메서드 생성 및 로직 변경
    - update와 달리 id외 추가 필드 정보가 필요로 하지 않다고 판단.

 
### 추가 반영사항
1. refactor: 각 Controller RequestMapping 적용
2. 인수테스트시 Map 대신 Request Dto를 통해 body 메세지를 던지도록 변경
3. 인수테스트 중복 코드를 메서드 추출하여 리팩토링

### 고민사항
1. section table이 가지는 up/donw_station_id에 대해 `Section` 도메인은 db와 일치하는 id(FK)값만 가질 것인가? 아니면 해당 id에 대한 Station객체를 직접 가질 것인가.
    - 객체를 직접 가지는 경우로 시도했으나, dao에서 사용되기 위해 getter에서 까지 반환해야하며, dao에서 조회시 rowMapper마다 타도메인인 Station을 의존해야하는 문제가 생겨서 순수 id만 가지도록 순회하였습니다. 관련 경험이나 지식이 부족해서 고민을 한 부분입니다.

2. 구간 변화시 매 구간이 변화할때마다 dao를 통해 생성/삭제/수정할 것인가? 아니면 Sections라는 일급컬렉션에서 모든 비지니스로직을 수행한 뒤 -> batchUpdate로 해결할 것인가.
    - 후자를 선택하였습니다. 하지만 중간역을 삭제하는 구간 삭제의 경우, seciton 2개가 삭제되고 1개가 생성되므로 section 1개가 실질적으로 dao를 통해 db에서 삭제되도록 하였습니다.
        - A-B-C 중 B를 삭제한다고 예를 들면
        - `A-B`  `B-C` 삭제 이후  `A-C`생성  이라고 생각하고
        -  `A-B`가 가졌던 sectionId를 `A-C` 구간 생성시 사용하게 했습니다.
        - 실제로 없어지는 구간인 `B-C` 구간의 경우 도메인 -> service로 반환하여 deleteById를 통해 해당 구간만 실질적으로 삭제하고, 나머지 구간은 `A-B` 에서 `A-C`로 수정하는 것으로 작성하였습니다. 더 많은 구간들이 존재하여 모든 구간을 batchUpdate하였습니다.
        - 잘 짜여진 구조라면 batchUpdate가 아니라 매번 생성/수정/삭제되는 구간들을 응답해서 dao로 생성/수정/삭제되도록 해야할 것 같은데, 쉽지가 않았습니다! **이 방식이 괜찮을까요?**

3. 테스트 격리
    - 인수테스트끼리는 DirtiesContext로 인해 격리가 잘되었습니다.
    - 하지만 Dao나 DomainTest에서도 영향을 미치는 것 같아, 생성후 삭제할 수 있으면 다 해줬습니다. **그래도 ㅠ_ㅜ 통과안되는 테스트들이 많아, 넣어주는 인자들을 변경(추가생성)해서 테스트를 통과**시켰습니다.
    - 범블비! 혹시 인수테스트/service테스트/dao테스트할 때 서로를 잘 격리시키기 위해 어떤  환경을 구성하는지 여쭈어봐도 되겠습니까!


4. 구간 생성/삭제에 대한 경우의 수를 enum으로 관리해보았습니다. 그간 미션에서 아래와 같이 정리가 되었었는데, 지금 적용해봐야할 것 같았습니다! 이것도 범블비 생각은 어떠하신지 궁금합니다!
   - 제한된 수의 응답을 해야하고 if문을 줄일 수 있다.
   - flag가 필요한 순간 boolean대신 사용하면 확장이 가능하다

is2js committed:
refactor: 불필요한 기본 생성자 삭제

is2js committed:
refactor: request/response DTO private 기본생성자 복구 및 package 추가

is2js committed:
refactor: LineAcceptanceTest requestBody 자료구조 변경(Map -> LineRequest)

is2js committed:
refactor: StationAcceptanceTest requestBody 자료구조 변경(Map -> StationRequest)

is2js committed:
refactor(test): 인수테스트시 중복되는 코드 메서드 추출

is2js committed:
refactor: 부적절한 Optional 사용 수정
 - rowMapper가 null 반환하지 않는 한, null응답의 가능성이 없는 queryForObject에 대해 Opiotnal.of()로 대체후 suppressWarning Method

is2js committed:
feat: Line name 검증 추가 및 테스트 작성

is2js committed:
feat: Station name 검증 추가 및 테스트 작성

is2js committed:
refactor: Dao delete(entity) 메소드들을 deleteById(id)로 변경

is2js committed:
feat: Line domain color에 대한 빈값 검증 추가

is2js committed:
refactor: 각 Controller RequestMapping 적용

is2js committed:
refactor: name에 대한 검증 Name 도메인에 위임

is2js committed:
feat: requestDto의 각 field에 대해 null 검증 추가

is2js committed:
docs: 3단계 요구사항 추가

is2js committed:
refactor(test): 테스드 코드 가독성을 위한 추가 리팩토링

is2js committed:
feat: Section 도메인 및 Test 코드 구현

is2js committed:
feat: Section 개별 조회 기능 구현

is2js committed:
feat: 특정 노선에 대한 Section 전체 조회 기능 구현

is2js committed:
feat: 특정 노선에 대한 모두 삭제 기능 구현

is2js committed:
feat: Line 및 LineRequest의 추가필드 정보 입력 가능하도록 수정

is2js committed:
feat: SectionService 생성

is2js committed:
feat: 정보 추가된 Line 생성 기능 및 Test 구현
- 반복되는 test 메서드 및 상수 Fixture 클래스로 추출

is2js committed:
fix: 불필요한 Line 필드 삭제

is2js committed:
feat: 정보 추가된 Line 전체 조회 기능 및 Test 구현

is2js committed:
feat: 정보 추가된 Line 개별 조회 기능 및 Test 구현

is2js committed:
refactor: 수정을 위한 LineRequest를 위해 section관련 필드 null허용

is2js committed:
feat: 정보 추가된 Line 수정 기능 및 Test 구현

is2js committed:
feat: 정보 추가된 Line 삭제 기능 및 Test 구현

is2js committed:
feat: Sections(구간정보) 생성 및 생성 검증 test 구현

is2js committed:
feat: sections 구간 등록시 상/하행 종점 비교 검증 추가 및 Test 구현

is2js committed:
feat: 구간 추가 경우의수를 나타낼 enum 추가

is2js committed:
feat: 구간을 중간에 추가시 거리 검증 기능 및 Test 구현

is2js committed:
feat: 중간역 등록시 기존 구간 제거후 쪼개진 구간들을 추가하는 기능 및 Test 구현

is2js committed:
feat: JdbcSectionDao 일괄 수정 기능 및 Test 구현

is2js committed:
refactor: section 정렬 순서 기준을 upStationId로 변경

is2js committed:
feat: 새로운 상/하행 종점 역 등록시 기존 구간에 추가하는 기능 및 Test 구현

is2js committed:
fix: 기존 구간정보에 여러 구간이 있어도 구간 추가 가능하도록 수정

is2js committed:
refactor: 구간 추가 경우의 수 판단 로직 간소화

is2js committed:
fix: 잘못된 메서드 호출 수정

is2js committed:
fix: 잘못된 메서드 호출 순서 수정

is2js committed:
feat: 구간 등록에 대한 인수테스트 작성

is2js committed:
docs: 구현 완료된 기능(구간 등록) 체크

is2js committed:
fix: JdbcSectionDao 데이터 저장 로직 오타 수정

is2js committed:
feat: 구간 등록에 대한 인수테스트 추가

is2js committed:
feat: 도메인에서 구간 삭제 검증 기능 및 Test 구현

is2js committed:
feat: 도메인에서 구간의 종점 삭제 기능 및 Test 구현

is2js committed:
feat: 도메인에서 구간의 중간역 삭제 기능 및 Test 구현

is2js committed:
refactor: Sections 및 SectionDeleteStatus 로직 간소화 리팩토링

is2js committed:
feat: SectionService에서 구간 삭제 기능 구현

is2js committed:
feat: 구간 삭제에 대한 인수테스트 추가

is2js committed:
docs(readme): 구간 제거 기능 구현 완료 체크

is2js committed:
feat: 노선별 남은 구간 자동 삭제를 위한 sql 제약조건 추가

is2js committed:
refactor: section 패키지 생성

is2js committed:
fix: test에 한하여 FK 제약조건 삭제

is2js committed:
fix: 인수테스트 에러 해결을 위한 테스트 코드 수정 및 불필요 파일 삭제

is2js committed:
feat: LineService에 대한 Test 구현

is2js committed:
feat: StationService에 대한 Test 구현

is2js committed:
fix: 통합테스트 격리를 위한 테스트 fixture 변경 사용

is2js committed:
style: 각 Test에 설명을 위한 주석 추가

is2js committed:
style: 코드 컨벤션 적용 및 불필요한 코드 삭제

ddaaac commented on src/main/java/wooteco/subway/domain/section/Section.java:
```diff
@@ -0,0 +1,138 @@
+        return Long.compare(section1.getUpStationId(), section2.getUpStationId());
+    }
+
+    public Section createMiddleToDownSection(final Section unSplitSection) {
```
로직은 getter 등보다 위쪽에 위치하게 해주세요.

is2js replied:
해당 부분 컨벤션 순서대로 수정하였습니다 감사합니다 범블비!

ddaaac commented on src/main/java/wooteco/subway/domain/section/Section.java:
```diff
@@ -0,0 +1,138 @@
+        return Long.compare(section1.getUpStationId(), section2.getUpStationId());
+    }
+
+    public Section createMiddleToDownSection(final Section unSplitSection) {
```
도메인 로직에 대한 테스트가 있으면 좋을 것 같네요 :)
현재 도메인에 대한 테스트가 빠진 부분이 많은데 꼼꼼히 채워주시면 좋을 것 같습니다.

is2js replied:
중요 로직인데 service test만 작성하였습니다! 도메인 테스트도 추가하였습니다! 
감사합니다 범블비!

ddaaac commented on src/main/java/wooteco/subway/domain/section/Section.java:
```diff
@@ -0,0 +1,138 @@
+        return downStationId;
+    }
+
+    @Override
+    public int compare(final Section section1, final Section section2) {
+        return Long.compare(section1.getUpStationId(), section2.getUpStationId());
+    }
```
Section이 comparable 을 구현하기 적당한 객체일까요?
upstationId를 기준으로 정렬되기 보다는 down과 up이 맞물리는 기준으로 정렬되어야할 것 같아요.
지금은 예를 들어 Section의 up = 3, down = 2라면 여러 버그들이 발생할 것 같습니다.

+) 정확히는 파악이 안되지만 아마도 이 compare는 사용되지 않는 것처럼도 보이구요.

is2js replied:
section은 상->하행의 순서가 stationId를 기준으로 이미 정해졌다고 생각했었습니다.
Comparable은 미사용되어 삭제했습니다(기존에는 stream에서 sorted()로 정렬해서 반환하도록 구현했었습니다!)

> upstationId를 기준으로 정렬되기 보다는 down과 up이 맞물리는 기준으로 정렬되어야할 것 같아요.

이 부분에 대해서 놓치고 있었습니다! stationId가 빠르면, 항상 상행 구간으로 생각했었네요ㅠ!

1. 구간 자신의 upStationId 한번도 전체 구간 중 downStationId로 사용되지 않을 경우, first station으로 판단
2. 이후 앞 구간의 downStationId와 같은 id를 upStationId를 가진 구간이 있을 경우, firstSation에 추가해서 연결하여 최종 반환


하도록 수정하였습니다. 감사합니다 범블비!

ddaaac commented on src/main/java/wooteco/subway/domain/Name.java:
```diff
@@ -0,0 +1,55 @@
+package wooteco.subway.domain;
+
+public class Name {
```
값 객체 구현 잘 해주셨습니다 💯

ddaaac commented on src/main/java/wooteco/subway/dao/JdbcSectionDao.java:
```diff
@@ -0,0 +1,91 @@
+        final String sql = "SELECT * FROM section WHERE id = :id";
+        final MapSqlParameterSource parameters = new MapSqlParameterSource("id", id);
+        try {
+            return Optional.of(namedParameterJdbcTemplate.queryForObject(sql, parameters, getSectionRowMapper()));
```
row mapper를 매번 생성할 필요는 없어보이네요 😄

is2js replied:
해당 부분을 static final로 변경하겠습니다!

ddaaac commented on src/main/java/wooteco/subway/domain/section/SectionAddStatus.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Objects;
+import java.util.function.BiPredicate;
+
+public enum SectionAddStatus {
```
> 구간 생성/삭제에 대한 경우의 수를 enum으로 관리해보았습니다. 그간 미션에서 아래와 같이 정리가 되었었는데, 지금 적용해봐야할 것 같았습니다! 이것도 범블비 생각은 어떠하신지 궁금합니다!

독특한 방식을 사용해주셨군요. 객체가 갖는 조건들을 포장하려는 시도가 좋네요 👍 

저라면 아마도 enum 대신 인터페이스를 사용했을 거 같습니다. 생각보다 enum의 가독성은 그렇게 좋은 편이 아니라고 생각하거든요. 확장할 때도 불편하구요.

또, 이왕에 이런 조건을 관리하는 객체를 만들었다면 좀 더 활용해볼 수 있을 거 같아요. 지금은 `Sections`에 아래 코드가 있는데요, 이 또한 여기로 옮겨올 수 있을 것 같네요.
```java
        if (hasMiddleSection(sectionAddStatus)) {
            addMiddleSection(section, sectionAddStatus);
            return getSortedByUpStationIdSections();
        }
        value.add(section);
```

+) 추가로, 이렇게 따로 조건을 관리하는 방법이 그냥 Sections에서 메서드 등으로 조건으로 관리하는 것에 비해 어떤 장단점이 있는지도 꼭 체크해보시면 좋겠어요. 솔직하게 말씀드리면 저는 다른 로직을 보다가 이 부분을 봐야할 때 로직을 파악하는 게 어렵다고 느껴지네요.

is2js replied:
enum 내부로 조건에 대한 로직만 넣어보았으나, 범블비께서 제안해주신 코드를 다 넣으면 원래 목적(조건에 대한 판단만 enum으로)과는 멀어지게, 구간을 add하는 로직까지 넘어가는 문제가 생겼습니다.

지난 미션들에서 interface로 추출후 외부에서 조건에 맞는 구현체를 통해 각 로직이 펼쳐지고 확장 또한 가능하다는 것을 경험해보았습니다.^^

다른 주제에 대해서도 충분히 연습해보고 enum을 활용에 대한 고민을 해보아야할 것 같습니다.ㅠ

감사합니다 범블비!

ddaaac replied:
넵넵 이 부분은 다양하게 시도해보시고 고민해보시면 좋을 거 같아요. 
정답이 있는 문제는 아닌 것 같네요 :)

ddaaac commented on src/main/java/wooteco/subway/domain/section/SectionAddStatus.java:
```diff
@@ -0,0 +1,68 @@
+            .anyMatch(it -> isOnlyUpStationSame(section, it.getUpStationId(), it.getDownStationId()));
+    }
+
+    private static boolean isOnlyUpStationSame(final Section section, final Long upStationId,
+                                               final Long downStationId) {
+        return Objects.equals(section.getUpStationId(), upStationId) &&
+            !Objects.equals(section.getDownStationId(), downStationId);
```
이런 로직은 Section#isOnlyUpStationSame(Section other)로 넣을 수 있지 않을까요?
getter가 등장하면 한 번쯤 도메인 내부로 로직을 넣을 수 있나 확인해주세요 :)

is2js replied:
네 범블비! stream내부라서 레벨1간 연습했던 내용을 적용못해버렸네요!
이번 기회로 다시 한번 상기시켜 이후 잘 적용하도록 해보겠습니다

감사합니다 범블비!

ddaaac commented on src/main/java/wooteco/subway/domain/section/SectionDeleteStatus.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    private static boolean isMiddleSection(final List<Section> sections, final Long stationId) {
+        if (isFirstSection(sections, stationId)) {
+            return false;
+        }
+        return !isLastSection(sections, stationId);
```
```suggestion
        return !isFirstSection(sections, stationId) && !isLastSection(sections, stationId);
```
이 부분은 취향차일 수 있으니 이렇게도 쓸 수 있다 정도만 알면 좋을 것 같네요. 이미 아실 수도 있구요 😄

is2js replied:
after 상태의 코드를 before 코드로 수정했었습니다.

최근에[ 분기문과 드모르간](https://blog.chojaeseong.com/java/if/conditionalstatements/a&&b/earlyreturn/boolean/%EC%9A%B0%ED%85%8C%EC%BD%94/2022/03/30/%EB%B6%84%EA%B8%B0%EB%AC%B8%EA%B3%BC-%EB%93%9C%EB%AA%A8%EB%A5%B4%EA%B0%84.html) 법칙에 대해 학습하고 적용하려 애쓰는 중이었습니다!
빨리 걸러지는 로직을 early return하면 효율상 이점을 가질 수 있고,  부정문도 줄어든다는 장점이 있었습니다!

범블비 생각은 어떠신지요?!

ddaaac replied:
저도 최근에 이런 스타일에 대해서 알게되었는데 깔끔하게 글로 정리해주셨네요. 감사합니다 🙇

ddaaac commented on src/main/java/wooteco/subway/domain/Station.java:
```diff
@@ -1,43 +1,49 @@
 package wooteco.subway.domain;
 
-import java.util.Objects;
+public class Station implements Comparable<Station> {
```
Station도 마찬가지로 정말 Comparable 인터페이스에 알맞은지 생각해보시면 좋을 거 같아요.
id는 식별자이지 크기의 비교대상이라고 보기에는 조금 어려울 것 같네요.

is2js replied:
네 범블비! id를 domain에서 사용하는 로직을 없애야할 것 같습니다!
좋은 의견 감사합니다 범블비!  해당 부분도 삭제하겠습니다!

ddaaac commented on src/main/java/wooteco/subway/service/SectionService.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @SuppressWarnings("OptionalGetWithoutIsPresent")
+    @Transactional
```
find도 트랜잭션의 대상일까요?

is2js replied:
이 부분에선 findAll이 아니라 불필요한 것 같습니다

감사합니다 범블비!

ddaaac commented on src/main/java/wooteco/subway/dto/request/LineRequest.java:
```diff
@@ -0,0 +1,73 @@
+//        Objects.requireNonNull(upStationId, ERROR_NULL);
+//        Objects.requireNonNull(downStationId, ERROR_NULL);
+//        Objects.requireNonNull(distance, ERROR_NULL);
+        // -> null 허용 + 수정용 필드 2개만 받는 생성자 추가 생성(수정시 id는 pathVariable로 옴)
```
주석은 지워주세요!

ddaaac commented on src/main/java/wooteco/subway/ui/LineController.java:
```diff
@@ -10,54 +10,84 @@
     }
 
-    @PostMapping("/lines")
+    @PostMapping()
```
불필요한 괄호는 제거해주세요.

ddaaac commented on src/main/java/wooteco/subway/service/SectionService.java:
```diff
@@ -0,0 +1,64 @@
+    public void addSection(final Long lineId, final SectionRequest sectionRequest) {
+        final List<Section> sectionsBeforeAddSection = sectionDao.findSectionStationsByLineId(lineId);
+        Section newSection = sectionRequest.toEntity(lineId);
+        newSection = sectionDao.save(newSection);
+        final List<Section> sectionsAfterAdd = new Sections(sectionsBeforeAddSection).addSection(newSection);
+        sectionDao.batchUpdate(sectionsAfterAdd);
```
> 잘 짜여진 구조라면 batchUpdate가 아니라 매번 생성/수정/삭제되는 구간들을 응답해서 dao로 생성/수정/삭제되도록 해야할 것 같은데, 쉽지가 않았습니다! 이 방식이 괜찮을까요?

지금 방식도 의의가 있다고 생각해요. 말씀해주신대로 메서드 시그니처들이 깔끔하게 뽑히거든요. 
다만, 불필요한 쿼리가 너무 많이 발생할 거 같네요. Section이 많으면 많을수록 더 심해질거구요.
조금 코드의 깔끔함을 희생하더라도 필요한 Section들만 저장/업데이트하면 좋을 것 같습니다.

is2js replied:
update가 필요한 section들만 추출하도록 수정하였습니다.

하지만, 많이 부족한 코드처럼 느껴집니다ㅠ^ㅠ

감사합니다 범블비!

ddaaac commented on src/test/resources/schema.sql:
```diff
@@ -1,24 +1,67 @@
-    id bigint auto_increment not null,
-    name varchar(255) not null unique,
-    primary key(id)
+    id
+    bigint
+    auto_increment
+    not
+    null,
+    name
+    varchar
+(
+    255
+) not null unique,
+    primary key
+(
+    id
+)
```
정렬이 이상해진 거 같아요!

is2js replied:
해당 부분 수정하였습니다!

ddaaac commented on src/main/java/wooteco/subway/domain/section/Section.java:
```diff
@@ -0,0 +1,138 @@
+    private static final String ERROR_SAME_STATION = "[ERROR] 상행 종점과 하행 종점이 같을 수 없습니다.";
+    private static final int INVALID_DISTANCE_STANDARD = 0;
+
+    private final Long id;
+    private final Long lineId;
+    private final int distance;
+    private final Long upStationId;
+    private final Long downStationId;
```
> section table이 가지는 up/donw_station_id에 대해 Section 도메인은 db와 일치하는 id(FK)값만 가질 것인가? 아니면 해당 id에 대한 Station객체를 직접 가질 것인가.

앞으로도 많은 고민을 하게 되실텐데요, 이렇게도 구현해보고 저렇게도 구현해보면서 직접 더 좋은 방법을 찾아가시면 될 거 같아요. 이런 부분에 대해서는 정답은 없거든요. 개인적으로 저는 id만 갖게 해도 충분하다고 생각합니다.

is2js replied:
네 범블비, id만 갖는 경우 이외에, 미션 복습할 때 다른 방식으로도 구현해보겠습니다.
아직 탐색하는 과정이라서 여러가지 시도해보고 있습니다

감사합니다 범블비!

ddaaac commented on src/test/java/wooteco/subway/dao/JdbcLineDaoTest.java:
```diff
@@ -31,63 +29,80 @@ void set() {
+        //then
+        assertThat(actualName).isEqualTo(LINE_2_GREEN.getName());
+
+        lineDao.deleteById(actual.getId());
```
> 혹시 인수테스트/service테스트/dao테스트할 때 서로를 잘 격리시키기 위해 어떤 환경을 구성하는지 여쭈어봐도 되겠습니까!

저도 이렇게 매번 지우는 방식을 선호합니다. 저희 환경에서는 테스트 데이터도 중요하고, 실제 DB와 유사한 환경에서 테스트하는 것도 중요하거든요. 그래서 약간의 불편함을 감수하고 이런 식으로 테스트합니다 :)

ddaaac reviewed (changes requested):
안녕하세요 돌범 ✋ 

3단계 구현 잘 해주셨어요! 

몇 가지 코멘트를 남겼으니 확인해주세요.

is2js committed:
refacotr: front test를 위한 prod db 설정 적용

is2js committed:
refactor: 메서드 호출 순서 컨벤션에 맞도록 수정

is2js committed:
fix(domain/Section): 불필요한 인터페이스 구현 및 메서드 삭제

is2js committed:
feat(test): domain Section에 대한 test 추가 구현

is2js committed:
fix: test를 위한 DDL 제약조건 제거

is2js committed:
refactor: sections 정렬 기준 변경

- 기존: 각 구간별 stationId로 정렬
- 변경:
    1. 구간 자신의 upStationId 한번도 전체 구간 중 downStationId로 사용되지 않을 경우, first station으로 판단
    2. 이후 앞 구간의 downStationId와 같은 id를 upStationId를 가진 구간이 있을 경우 추가해서 연결

is2js committed:
refactor: rowMapper static final로 변경

is2js committed:
refactor: 일부 로직 간소화

is2js committed:
refactor: 같은 domain간 비교로직 domain내부로 로직 이동

is2js committed:
refactor(domain/Station): 불필요한 인터페이스 구현 삭제

is2js committed:
refactor: find 로직에 불필요한 transactional 애너테이션 삭제

is2js committed:
chore: 불필요한 주석 및 괄호 삭제

is2js committed:
feat: 구간 등록 및 삭제 로직 변경(batchUpdate -> 단일 CRUD)

is2js commented:
### 추가 반영사항

1. front test를 위한 prod db 설정 적용 
2. 메서드 호출순서 수정
3. fix(domain/Section): 불필요한 인터페이스 구현 및 메서드 삭제
4. domain Section에 대한 test 추가 구현

5. section 정렬에 대한 기준 변경
	1. sattion 순서대로만 연결: 1-2, 2-3, 3-4 -> upstationId로 정렬가능
	2. sattion 순서와 무관하게 연결: 4-1, 1-2, 2-3 -> downStation에 포함되지 않는 station이 firstStation(noneMatch)
	    - firstStation을 찾고 -> 거기서부터 하나씩 연결하여 반환하기
		1. 구간 자신의 upStationId 한번도 전체 구간 중 downStationId로 사용되지 않을 경우, first station으로 판단
		2. 이후 앞 구간의 downStationId와 같은 id를 upStationId를 가진 구간이 있을 경우 추가해서 연결하여 최종 반환
		
- findFirst로 반환하여 -> 밖에서 .isPresent()로 while문을 돈다... 그 find조건은 마지막에 업데이트한다. ->
  
  
6. section vs section의 연산일때, get 대신 메세지 보내기!
	- 같은 domain간 비교로직 domain내부로 로직 이동
	
7. find 로직에 불필요한 transactional 애너테이션 삭제
8. batchUpdate는 메서드 시그니쳐가 깔끔하나, 개별 수정/삭제해야.. 나중에 section이 많아질때를 대비해서 처리할 수 있다.
	- 100만개의 section이 있는 section라면.. 99만개이상이 업데이트..
	1. 구간 등록시 add완료된 sections 반환하여 batchUpdate -> 구간 save이후 기존 구간을 쪼갠 구간으로 update 필요한 경우(중간역 추가), optional이용하여 확인 후, 단일 update 
	2. 구간 삭제시 remove완료된 sections 반환 batchUpdate -> 구간 delete이후 기존 구간 update가 필요한 경우(중간역 삭제), flag로 확인 후 단일 update
	
	```java
	@Transactional
	public void addSection(final Long lineId, final SectionRequest sectionRequest) {
		final List<Section> sectionsBeforeAddSection = sectionDao.findSectionStationsByLineId(lineId);
		Section newSection = sectionRequest.toEntity(lineId);
		newSection = sectionDao.save(newSection);
		new Sections(sectionsBeforeAddSection).addSection(newSection)
			.ifPresent(sectionDao::update);
	}

	@Transactional
	public void deleteSection(final Long id, final Long stationId) {
		lineDao.findById(id)
			.orElseThrow(() -> new LineNotFoundException("[ERROR] 해당 노선이 없습니다."));
		stationDao.findById(id)
			.orElseThrow(() -> new StationNotFoundException("[ERROR] 해당 이름의 지하철역이 존재하지 않습니다."));

		final Sections sections = new Sections(sectionDao.findSectionStationsByLineId(id));
		final boolean isMiddleDelete = sections.isMiddleDelete(stationId);
		sectionDao.deleteById(sections.deleteSectionByStationId(stationId));
		if (isMiddleDelete) {
			final Section updatedSection = sections.getUpdatedSection(
				new Sections(sectionDao.findSectionStationsByLineId(id)));
			sectionDao.update(updatedSection);
		}
	}
	```

범블비 덕분에 기본적인 것부터 생소한 것까지 다 다루는 미션이 되었던 것 같습니다!
가르침 주신 것에 대해 잘 기록해놓겠습니다!

ddaaac commented on src/main/java/wooteco/subway/domain/section/Section.java:
```diff
@@ -0,0 +1,150 @@
+        return downStationId;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        final Section section = (Section) o;
+
+        if (getDistance() != section.getDistance()) {
+            return false;
+        }
+        if (getId() != null ? !getId().equals(section.getId()) : section.getId() != null) {
+            return false;
```
equals & hashcode는 특별히 재정의할 일이 없으면 IDE에서 만들어주는대로 사용해도 될 거 같아요.
id가 있는 객체의 동등성을 판단한다면 id만 비교해도 좋을 거 같구요.

is2js replied:
ide 가 결정해주는 것인데, 음.. 정의 옵션에 gradle -> intellij로 선택하는 사항이 있길래 바꿨더니 코드스타일이 약간 달라진 것 같아요. 오호호.. 협업을 위해서 default로 바꿔놔야겠습니다^^ 

감사합니다 범블비!

ddaaac commented on src/main/java/wooteco/subway/domain/section/Sections.java:
```diff
@@ -0,0 +1,226 @@
+        return section.createUpToMiddleSection(sameDownStationSection);
+    }
+
+    public List<Section> getSortedSections() {
```
여기도 테스트가 필요할 거 같아요!

is2js replied:
private메서드인데, test를 위해 public으로 돌리고, 테스트코드로 작동확인후 테스트 코드는 삭제했었습니다.

private으로 다시 안돌려놨네요 흠..^^ 이것 역시 잘 기억해두도록 한번 기록을 해두겠습니다^^

감사합니다 범블비!

ddaaac commented on src/main/java/wooteco/subway/domain/section/Sections.java:
```diff
@@ -0,0 +1,226 @@
+
+    private Optional<Section> findNextSection(final Section previousSection) {
+        return value.stream()
+            .filter(section -> section.getUpStationId().equals(previousSection.getDownStationId()))
```
리뷰를 거치며 Section내에 추가된 메서드들을 활용할 수 있겠네요. 이렇게 도메인 내부로 메서드를 넣고 나서는 중복되는 메서드들이 있는지 꼭 확인해주세요!

is2js replied:
이 부분 역시 배우네요. 새로운 메서드가 추가될 경우, 이전 코드에서 반영할지도 한번 살펴보겠습니다.
아직 주니어도 못된 취준생이라 배울 것이 많았는데, 이런 의식적인 것들을 하나하나 배우니 좋습니다. 잘 기록했다가 다음에 잘 반영되도록 하겠습니다^^

ddaaac reviewed (approved):
안녕하세요 돌범!

리뷰 반영 잘 해주셨습니다 👍  여기서 머지해도 될 거 같아요.
마지막 리뷰가 될 거 같아서 아쉽네요 😭 

간단한 코멘트 몇 가지 남겼으니 확인해주세요.
앞으로도 화이팅입니다 💪
